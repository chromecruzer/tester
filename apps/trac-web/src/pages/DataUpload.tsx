import React, {useEffect, useState} from 'react';
import _ from 'lodash';
import {useNavigate, useParams} from "react-router";
import useWebSocket from "react-use-websocket";
import PagesMap from "../navigation/PagesMap";
import {Outlet} from "react-router-dom";
import {MdOutlineCancel, MdOutlinePublishedWithChanges} from "react-icons/md";
import {connectionStatus, DataUploadTypes, notifyPaths, NullableString, UploadSessionToken} from "@trac/datatypes";
import {useMutation, useQueryClient} from "@tanstack/react-query";
import {postCancel, postCommit} from "../api/uploadApi";
import {useTracContext} from "../TracContext";
import {wsRoot} from "../httpConfig";
import {UploadStatus} from "../components/upload/UploadStatus";

const pagesMap = new PagesMap();
const DataUpload = () => {
  const navigate = useNavigate();
  const queryClient = useQueryClient();
  const {session: sessionFromParams, upload_filetype} = useParams();
  const [error, setError] = useState(null as NullableString);
  const [isLoading, setIsLoading] = useState(false);
  const {getTracContext, setTracContext} = useTracContext();
  const uploadSession = getTracContext('uploadSession') as UploadSessionToken || null;
  const session = sessionFromParams || null;
  const uploadFileType = (upload_filetype || null) as (DataUploadTypes | null);
  // console.log(`websocket URL ${wsRoot}${notifyPaths['iolUpload']}`);

  const {lastMessage, readyState } = useWebSocket(`${wsRoot}${notifyPaths['iolUpload']}`,
    {share: true, retryOnError: true});
  console.log('websocket returned data', connectionStatus[readyState], lastMessage)
  const cancelMutation = useMutation({
    mutationFn: async () => {
      setIsLoading(true);
      setError(null);
      return postCancel(session, uploadFileType);
    },
    onSuccess: (data, variables, context) => {
      setError(null);
      setTracContext('uploadSession', null);
      queryClient.invalidateQueries();
    },
    onError: (error:NullableString, variables, context) => {
      if(error) {
        setError(error);
      }
    },
    onSettled: (data, error, variables, context) => {
      setIsLoading(false);
      navigate(`/${pagesMap.page('Data Upload')}`, {replace: true});
    }
  })
  const commitMutation = useMutation({
    mutationFn: async () => {
      setIsLoading(true);
      setError(null);
      return postCommit(session, uploadFileType);
    },
    onSuccess: (data, variables, context) => {
      setError(null);
      setTracContext('uploadSession', null);
    },
    onError: (error:NullableString, variables, context) => {
      if(error) {
        setError(error);
      }
    },
    onSettled: (data, error, variables, context) => {
      setIsLoading(false);
      queryClient.removeQueries({queryKey:['data history']})
      navigate(`/${pagesMap.page('Data Upload')}`);
    }
  })
  useEffect(() => {
  //   console.log(`sessions and file types
  // from parameters ${session} ${upload_filetype}
  // from redux "${uploadSession?.session}" "${uploadSession?.type}"`);
    // After submitting an upload, the session and filetype shows up in the context instead of the parameters.
    if ((!_.isString(session)) && uploadSession &&
      _.isString(uploadSession.session) && _.isString(uploadSession.type)) {
      // console.log('navigating to Analytics');
      navigate(pagesMap.page('Analytics', [uploadSession.type, uploadSession.session]))
    }

  }, [uploadSession]);
  const handleCancel = async () => {
    if (session && uploadFileType) {
      cancelMutation.mutate();
    }
  }

  const handleCommit = async () => {
    if (session && upload_filetype) {
      commitMutation.mutate();
    }
  }
  console.log(`For the button bar ${session} and ${upload_filetype}`);
  const buttonBar = session && upload_filetype &&
    <div className='flex gap-4 items-center'>
      <button
        onClick={handleCancel}
        className='button button-enabled-outlined'
      ><MdOutlineCancel/>Cancel
      </button>
      <button
        onClick={handleCommit}
        className={`${isLoading ? 'button-disabled' : 'button-enabled'} button`}>
        <MdOutlinePublishedWithChanges/>{isLoading ? 'Committing' : 'Commit'}
      </button>
    </div>;
  const errorMessage = error && <div><h4>Network Error</h4></div>;
  const loadingMessage = isLoading && <UploadStatus
    readyStatus={readyState}
    message={lastMessage}/>

  return <div>
    {buttonBar}
    {loadingMessage}
    {errorMessage}
    <Outlet/>
  </div>;
}
export default DataUpload;
