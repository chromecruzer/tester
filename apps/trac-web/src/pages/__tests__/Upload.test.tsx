import React from "react";
import DataUpload, {pagesMap} from "../DataUpload";
import * as reactRedux from "../../redux/hooks";
import * as reactRouter from "react-router";
import * as reduxActions from '../../redux/uploadActions'
import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
jest.mock('../../components/Upload/UploadSubmit', () => ({
  UploadSubmit: props => (<>
    <h4>Mock Submit</h4>
  </>),
}))

jest.mock('../../components/Upload/UploadDetailsComponent', () => ({
  UploadComparisonsComponent: props => (<>
    <h4>Mock Comparisons</h4>
  </>),
}))

jest.mock('../../components/Upload/UploadAnalyticsComponent', () => ({
  UploadAnalyticsComponent: props => (<>
    <h4>Mock Analytics</h4>
  </>),
}))


describe('Upload', () => {
  let dispatchSpy, cancelReduxMock, commitReduxMock, navigateSpy, useParamsMock, useSelectorMock, activePageMock;
  beforeEach(() => {
    navigateSpy = jest.fn();
    dispatchSpy = jest.fn().mockReturnValue(Promise.resolve());
    useParamsMock = jest.spyOn(reactRouter, 'useParams');
    jest.spyOn(reactRouter, 'useNavigate').mockImplementation(() => navigateSpy);
    jest.spyOn(reactRedux, 'useAppDispatch').mockImplementation(() => dispatchSpy);
    commitReduxMock = jest.spyOn(reduxActions, 'postCommit')
    cancelReduxMock = jest.spyOn(reduxActions, 'postCancel')
    activePageMock = jest.spyOn(pagesMap, 'activePage');
    useSelectorMock = jest.spyOn(reactRedux, 'useAppSelector');
    useSelectorMock.mockImplementation(() => ({}));
  });
  const renderUploadPage = () => render(
    <reactRouter.MemoryRouter><DataUpload/></reactRouter.MemoryRouter>);
  it('should display an http error', () => {
    const expected = {error:'we have an error'};
    useParamsMock.mockImplementation(() => ({upload_filetype: 'iol_record', session: 'SESSION'}));
    useSelectorMock.mockImplementation(() => expected);
    renderUploadPage();
    screen.getByText(expected.error)
  });
  it('should have a button that does a cancel', async () => {
    useParamsMock.mockImplementation(() => ({upload_filetype: 'iol_record', session: 'SESSION'}));
    renderUploadPage();
    const button = screen.queryByRole('button', {name: 'Cancel'});
    if (button === null) {
      throw new Error('Cannot find cancel button');
    }
    return userEvent.click(button).then(() => {
      expect(cancelReduxMock.mock.calls).toMatchSnapshot();
    });
  });
  it('should have a button that should go to details', () => {
    useParamsMock.mockImplementation(() => ({upload_filetype: 'iol_record', session: 'SESSION'}));
    activePageMock.mockImplementation(() => 'Analytics');
    renderUploadPage();
    const button = screen.queryByRole('button', {name: 'Details'});
    if (button === null) {
      throw new Error('Cannot find details button');
    }
    return userEvent.click(button).then(() => {
      expect(navigateSpy.mock.calls).toMatchSnapshot();
    });
  });
  it('should have a button that does a commit', async () => {
    useParamsMock.mockImplementation(() => ({upload_filetype: 'iol_record', session: 'SESSION'}));
    renderUploadPage();
    const button = screen.queryByRole('button', {name: 'Commit'});
    if (button === null) {
      throw new Error('Cannot find commit button');
    }
    userEvent.click(button).then(() => {
      expect(commitReduxMock.mock.calls).toMatchSnapshot();
    });
  });
});
