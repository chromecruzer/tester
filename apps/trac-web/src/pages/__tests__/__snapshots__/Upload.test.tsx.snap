// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataUpload should have a button that does a cancel 1`] = `
Array [
  Array [
    "SESSION",
    "iol_record",
  ],
]
`;

exports[`DataUpload should have a button that should go to details 1`] = `
Array [
  Array [
    "details/iol_record/SESSION",
  ],
]
`;
